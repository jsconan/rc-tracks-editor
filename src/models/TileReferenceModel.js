/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import getTileComponent from '../helpers/getTileComponent.js';
import getTileModel from '../helpers/getTileModel.js';
import { isDirectionValid, isTypeValid, STRAIGHT_TILE_TYPE, TILE_DIRECTION_RIGHT } from '../helpers/types.js';
import uid from '../helpers/uid.js';

/**
 * @typedef {import('./TileModel')} TileModel
 */

/**
 * @typedef {import('svelte').SvelteComponent} SvelteComponent
 */

/**
 * Represents a reference to a tile model.
 */
export default class TileReferenceModel {
    /**
     * Represents a reference to a tile model with the given specifications.
     * @param {string} type - The type of referenced tile.
     * @param {string} direction - The direction of the tile, it can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     */
    constructor(type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        this.id = uid();
        this.setType(type);
        this.setDirection(direction);
        this.setRatio(ratio);
    }

    /**
     * Sets the type of referenced tile.
     * @param {string} type - The type of referenced tile.
     * @returns {TileReferenceModel} - Chains the instance.
     * @throws {TypeError} - If the given type is not valid.
     */
    setType(type) {
        if (!isTypeValid(type)) {
            throw new TypeError('A valid type of tile is needed!');
        }

        this.type = type;
        return this;
    }

    /**
     * Sets the direction of the tile.
     * It can be either TileModel.DIRECTION_RIGHT or TileModel.DIRECTION_LEFT.
     * @param {string} direction - The direction of the tile.
     * @returns {TileReferenceModel} - Chains the instance.
     * @throws {TypeError} - If the given direction is not valid.
     */
    setDirection(direction) {
        if (!isDirectionValid(direction)) {
            throw new TypeError('A valid direction is needed!');
        }

        this.direction = direction;
        return this;
    }

    /**
     * Sets the size factor of the tile.
     * 1 means the tile fits 1 tile section in each direction.
     * @param {number} ratio - The size factor of the tile.
     * @returns {TileReferenceModel} - Chains the instance.
     */
    setRatio(ratio) {
        this.ratio = Math.abs(ratio || 1);
        return this;
    }

    /**
     * Creates the model of the tile with the given size constraints with respect to the stored reference.
     * @param {number} laneWidth - The width of the track lane (the distance between the barriers).
     * @param {number} barrierWidth - The width of the barriers.
     * @param {number} barrierChunks - The number of barrier chunks per section.
     * @returns {TileModel} - Returns a tile model of the expected type.
     */
    getModel(laneWidth, barrierWidth, barrierChunks) {
        const Model = getTileModel(this.type);
        // @ts-expect-error
        return new Model(laneWidth, barrierWidth, barrierChunks, this.ratio);
    }

    /**
     * Gets the component constructor of the tile with respect to the stored reference.
     * @returns {SvelteComponent} - Returns the constructor of the component.
     */
    getComponent() {
        return getTileComponent(this.type);
    }
}
