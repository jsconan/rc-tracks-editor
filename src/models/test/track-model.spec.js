/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import CurvedTileModel from '../curved-tile-model.js';
import TileModel from '../tile-model.js';
import TileReferenceModel from '../tile-reference-model.js';
import TrackModel from '../track-model.js';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;

describe('TrackModel', () => {
    it('is a class', () => {
        expect(TrackModel).toEqual(expect.any(Function));
    });

    it('can get the index of a tile in the track', () => {
        const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

        track.addTile();
        const id = track.addTile();
        track.addTile();

        expect(track.getTileIndex(id)).toBe(1);
        expect(track.getTileIndex('id')).toBe(-1);
    });

    describe('can get a tile from the track', () => {
        it('by its identifier', () => {
            const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

            track.addTile();
            const id = track.addTile();
            track.addTile();

            expect(track.getTile(id)).toBe(track.track[1]);
            expect(track.getTile(id)).toBeInstanceOf(TileReferenceModel);
            expect(track.getTile('id')).toBeNull();
        });

        it('by its index', () => {
            const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

            track.addTile();
            track.addTile();
            track.addTile();

            expect(track.getTileAt(1)).toBe(track.track[1]);
            expect(track.getTileAt(1)).toBeInstanceOf(TileReferenceModel);
            expect(track.getTileAt(3)).toBeNull();
        });
    });

    describe('can build a track', () => {
        it('with the given size', () => {
            const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

            expect(track).toBeInstanceOf(TrackModel);
            expect(track).toMatchSnapshot();
        });

        describe('adding tiles', () => {
            describe('at the last position', () => {
                it('with the default specifications', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    const id = track.addTile();

                    expect(id).toEqual(expect.any(String));
                    expect(track).toMatchSnapshot();
                });

                it('with a particular type', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    const id = track.addTile(CurvedTileModel.TYPE, TileModel.DIRECTION_LEFT, 2);

                    expect(id).toEqual(expect.any(String));
                    expect(track).toMatchSnapshot();
                });
            });

            describe('at the first position', () => {
                it('with the default specifications', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();
                    const id = track.addFirstTile();

                    expect(id).toEqual(expect.any(String));
                    expect(track).toMatchSnapshot();
                });

                it('with a particular type', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();
                    const id = track.addFirstTile(CurvedTileModel.TYPE, TileModel.DIRECTION_LEFT, 2);

                    expect(id).toEqual(expect.any(String));
                    expect(track).toMatchSnapshot();
                });
            });
        });

        describe('removing tiles', () => {
            it('from the start', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                const id = track.addTile();
                track.addTile();
                track.addTile();

                expect(track.removeTile(id)).toBeTruthy();
                expect(track).toMatchSnapshot();
            });

            it('from the middle', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();
                const id = track.addTile();
                track.addTile();

                expect(track.removeTile(id)).toBeTruthy();
                expect(track).toMatchSnapshot();
            });

            it('from the end', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();
                track.addTile();
                const id = track.addTile();

                expect(track.removeTile(id)).toBeTruthy();
                expect(track).toMatchSnapshot();
            });

            it('at inexistent position', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();

                expect(track.removeTile('id')).toBeFalsy();
                expect(track).toMatchSnapshot();
            });
        });

        describe('replacing tiles', () => {
            it('with a tile having default specifications', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();
                const id = track.addTile();
                track.addTile();
                const newId = track.replaceTile(id);

                expect(newId).toEqual(expect.any(String));
                expect(newId).not.toBe(id);
                expect(track).toMatchSnapshot();
            });

            it('with a tile having a particular type', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();
                track.addTile();
                const id = track.addTile();
                const newId = track.replaceTile(id, CurvedTileModel.TYPE, TileModel.DIRECTION_LEFT, 2);

                expect(newId).toEqual(expect.any(String));
                expect(newId).not.toBe(id);
                expect(track).toMatchSnapshot();
            });

            it('at inexistent position', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();
                expect(track.replaceTile('id')).toBeNull();
                expect(track).toMatchSnapshot();
            });
        });

        describe('inserting tiles', () => {
            describe('before', () => {
                it('an inexistent position', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();

                    expect(track.insertBefore('id')).toBeNull();
                    expect(track).toMatchSnapshot();
                });

                it('the first position', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    const id = track.addTile();
                    track.addTile();
                    const newId = track.insertBefore(id);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });

                it('the last position', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();
                    const id = track.addTile();
                    const newId = track.insertBefore(id);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });

                it('a position in the middle', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();
                    const id = track.addTile();
                    track.addTile();
                    const newId = track.insertBefore(id);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });

                it('with a particular type', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    const id = track.addTile();
                    const newId = track.insertBefore(id, CurvedTileModel.TYPE, TileModel.DIRECTION_LEFT, 2);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });
            });

            describe('after', () => {
                it('an inexistent position', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();

                    expect(track.insertAfter('id')).toBeNull();
                    expect(track).toMatchSnapshot();
                });

                it('the first position', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    const id = track.addTile();
                    track.addTile();
                    const newId = track.insertAfter(id);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });

                it('the last position', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();
                    const id = track.addTile();
                    const newId = track.insertAfter(id);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });

                it('a position in the middle', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    track.addTile();
                    const id = track.addTile();
                    track.addTile();
                    const newId = track.insertAfter(id);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });

                it('with a particular type', () => {
                    const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                    const id = track.addTile();
                    const newId = track.insertAfter(id, CurvedTileModel.TYPE, TileModel.DIRECTION_LEFT, 2);

                    expect(newId).toEqual(expect.any(String));
                    expect(newId).not.toBe(id);
                    expect(track).toMatchSnapshot();
                });
            });
        });

        describe('for rendering', () => {
            it('using default position', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();
                track.addTile(CurvedTileModel.TYPE);
                track.addTile();
                track.addTile(CurvedTileModel.TYPE);
                track.addTile();
                track.addTile(CurvedTileModel.TYPE);
                track.addTile();
                track.addTile(CurvedTileModel.TYPE);

                expect(track.build()).toMatchSnapshot();
            });

            it('using a start position and angle', () => {
                const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);

                track.addTile();
                track.addTile(CurvedTileModel.TYPE);
                track.addTile();
                track.addTile(CurvedTileModel.TYPE);
                track.addTile();
                track.addTile(CurvedTileModel.TYPE);
                track.addTile();
                track.addTile(CurvedTileModel.TYPE);

                expect(track.build(100, 100, 45)).toMatchSnapshot();
            });
        });
    });

    it('can export to an object', () => {
        const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);
        track.addTile(CurvedTileModel.TYPE);
        expect(track.export()).toMatchSnapshot();
    });

    it('can import from an object', () => {
        const track = new TrackModel(laneWidth, barrierWidth, barrierChunks);
        const data = [
            {
                type: CurvedTileModel.TYPE,
                direction: TileModel.DIRECTION_LEFT,
                ratio: 2
            }
        ];

        track.addTile();

        expect(track.import({})).toBe(track);
        expect(track).toMatchSnapshot();

        expect(track.import(data)).toBe(track);
        expect(track).toMatchSnapshot();
    });
});
