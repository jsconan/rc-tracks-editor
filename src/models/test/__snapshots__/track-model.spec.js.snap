// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackModel can build a track adding tiles at the first position with a particular type 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 1,
      "id": "tile-14",
      "ratio": 2,
      "type": "curved-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-13",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track adding tiles at the first position with the default specifications 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-12",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-11",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track adding tiles at the last position with a particular type 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 1,
      "id": "tile-10",
      "ratio": 2,
      "type": "curved-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track adding tiles at the last position with the default specifications 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-9",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track for rendering using a start position and angle 1`] = `
Object {
  "height": 421.1269837220809,
  "tiles": Array [
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-68",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 45,
      "type": "straight-tile",
      "x": 100,
      "y": 100,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-69",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 45,
      "type": "curved-tile",
      "x": 22.218254069479784,
      "y": 177.78174593052023,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-70",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 135,
      "type": "straight-tile",
      "x": -55.56349186104045,
      "y": 177.78174593052023,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-71",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 135,
      "type": "curved-tile",
      "x": -133.3452377915607,
      "y": 100.00000000000001,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-72",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 225,
      "type": "straight-tile",
      "x": -133.3452377915607,
      "y": 22.218254069479798,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-73",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 225,
      "type": "curved-tile",
      "x": -55.563491861040475,
      "y": -55.56349186104045,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-74",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 315,
      "type": "straight-tile",
      "x": 22.218254069479755,
      "y": -55.56349186104046,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-75",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 315,
      "type": "curved-tile",
      "x": 100,
      "y": 22.21825406947974,
    },
  ],
  "width": 421.1269837220809,
  "x": -227.2361107568208,
  "y": -149.45436482630055,
}
`;

exports[`TrackModel can build a track for rendering using default position 1`] = `
Object {
  "height": 330,
  "tiles": Array [
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-60",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 0,
      "type": "straight-tile",
      "x": 0,
      "y": 0,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-61",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 0,
      "type": "curved-tile",
      "x": 0,
      "y": 110,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-62",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 90,
      "type": "straight-tile",
      "x": -55,
      "y": 165,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-63",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 90,
      "type": "curved-tile",
      "x": -165,
      "y": 165,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-64",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 180,
      "type": "straight-tile",
      "x": -220,
      "y": 110,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-65",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 180,
      "type": "curved-tile",
      "x": -220,
      "y": 0,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-66",
      "model": StraightTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 270,
      "type": "straight-tile",
      "x": -165,
      "y": -55.00000000000001,
    },
    Object {
      "component": [Function],
      "direction": 0,
      "id": "tile-67",
      "model": CurvedTileModel {
        "barrierChunks": 4,
        "barrierWidth": 5,
        "laneWidth": 80,
        "length": 110,
        "padding": 10,
        "ratio": 1,
        "width": 90,
      },
      "ratio": 1,
      "rotation": 270,
      "type": "curved-tile",
      "x": -55,
      "y": -55.00000000000003,
    },
  ],
  "width": 330,
  "x": -275,
  "y": -110.00000000000003,
}
`;

exports[`TrackModel can build a track inserting tiles after a position in the middle 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-54",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-55",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-57",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-56",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles after an inexistent position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-47",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles after the first position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-48",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-50",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-49",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles after the last position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-51",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-52",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-53",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles after with a particular type 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-58",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 1,
      "id": "tile-59",
      "ratio": 2,
      "type": "curved-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles before a position in the middle 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-41",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-44",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-42",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-43",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles before an inexistent position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-34",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles before the first position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-37",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-35",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-36",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles before the last position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-38",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-40",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-39",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track inserting tiles before with a particular type 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 1,
      "id": "tile-46",
      "ratio": 2,
      "type": "curved-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-45",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track removing tiles at inexistent position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-24",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track removing tiles from the end 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-21",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-22",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track removing tiles from the middle 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-18",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-20",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track removing tiles from the start 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-16",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-17",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track replacing tiles at inexistent position 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-33",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track replacing tiles with a tile having a particular type 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-29",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-30",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 1,
      "id": "tile-32",
      "ratio": 2,
      "type": "curved-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track replacing tiles with a tile having default specifications 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-25",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-28",
      "ratio": 1,
      "type": "straight-tile",
    },
    TileReferenceModel {
      "direction": 0,
      "id": "tile-27",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can build a track with the given size 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [],
}
`;

exports[`TrackModel can export to an object 1`] = `
Array [
  Object {
    "direction": 0,
    "ratio": 1,
    "type": "curved-tile",
  },
]
`;

exports[`TrackModel can import from an object 1`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 0,
      "id": "tile-77",
      "ratio": 1,
      "type": "straight-tile",
    },
  ],
}
`;

exports[`TrackModel can import from an object 2`] = `
TrackModel {
  "barrierChunks": 4,
  "barrierWidth": 5,
  "laneWidth": 80,
  "track": Array [
    TileReferenceModel {
      "direction": 1,
      "id": "tile-78",
      "ratio": 2,
      "type": "curved-tile",
    },
  ],
}
`;
